File: JokeGenerator.java
// JokeGenerator.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

class Joke {
    int id;
    String text;
    String category;

    Joke(int id, String text, String category) {
        this.id = id;
        this.text = text;
        this.category = category;
    }
}

public class JokeGenerator {
    private List<Joke> jokes = new ArrayList<>();
    private Random rand = new Random();

    public JokeGenerator() {
        loadDefaultJokes();
    }

    private void loadDefaultJokes() {
        jokes.add(new Joke(1, "Why don't scientists trust atoms? Because they make up everything!", "science"));
        jokes.add(new Joke(2, "What do you call a fake noodle? An impasta!", "food"));
        jokes.add(new Joke(3, "Why did the scarecrow win an award? Because he was outstanding in his field!", "pun"));
        jokes.add(new Joke(4, "I'm reading a book on anti-gravity. It's impossible to put down!", "book"));
        jokes.add(new Joke(5, "What's the best thing about Switzerland? I don't know, but the flag is a big plus!", "geography"));
        jokes.add(new Joke(6, "Why do programmers prefer dark mode? Because light attracts bugs!", "programming"));
        jokes.add(new Joke(7, "I told my wife she was drawing her eyebrows too high. She looked surprised.", "marriage"));
        jokes.add(new Joke(8, "What do you call a bear with no teeth? A gummy bear!", "animal"));
        jokes.add(new Joke(9, "Why don't eggs tell jokes? They'd crack each other up!", "food"));
        jokes.add(new Joke(10, "I used to play piano by ear, but now I use my hands.", "music"));
    }

    public Joke getRandomJoke() {
        if (jokes.isEmpty()) return null;
        return jokes.get(rand.nextInt(jokes.size()));
    }

    public List<Joke> getAllJokes() { return jokes; }

    public Joke addJoke(String text, String category) {
        int id = jokes.stream().mapToInt(j -> j.id).max().orElse(0) + 1;
        if (category == null || category.trim().isEmpty()) category = "general";
        Joke joke = new Joke(id, text, category);
        jokes.add(joke);
        return joke;
    }

    public boolean removeJoke(int id) {
        return jokes.removeIf(j -> j.id == id);
    }

    public List<Joke> getByCategory(String category) {
        return jokes.stream().filter(j -> j.category.equalsIgnoreCase(category)).collect(Collectors.toList());
    }

    public List<String> getCategories() {
        return jokes.stream().map(j -> j.category).distinct().sorted().collect(Collectors.toList());
    }

    public void saveToFile(String filename) throws IOException {
        String json = jokes.stream().map(j -> String.format("{\"id\":%d,\"text\":\"%s\",\"category\":\"%s\"}", j.id, j.text, j.category))
                .collect(Collectors.joining(",", "[", "]"));
        Files.write(Paths.get(filename), json.getBytes());
    }

    public void loadFromFile(String filename) throws IOException {
        String json = new String(Files.readAllBytes(Paths.get(filename)));
        // Simple parsing (not using a JSON lib to keep dependency-free)
        jokes.clear();
        String[] items = json.substring(1, json.length()-1).split("\\},\\{");
        for (String item : items) {
            String clean = item.replace("{", "").replace("}", "");
            String[] fields = clean.split(",");
            int id = 0;
            String text = "", category = "";
            for (String f : fields) {
                String[] kv = f.split(":");
                if (kv.length == 2) {
                    String key = kv[0].trim().replace("\"", "");
                    String val = kv[1].trim().replace("\"", "");
                    if (key.equals("id")) id = Integer.parseInt(val);
                    else if (key.equals("text")) text = val;
                    else if (key.equals("category")) category = val;
                }
            }
            jokes.add(new Joke(id, text, category));
        }
    }

    private static void displayJoke(Joke joke) {
        if (joke == null) {
            System.out.println("No jokes available.");
            return;
        }
        System.out.printf("\n😂 Joke: %s\n", joke.text);
        if (joke.category != null && !joke.category.isEmpty())
            System.out.printf("📁 Category: %s\n", joke.category);
    }

    public static void main(String[] args) throws IOException {
        JokeGenerator gen = new JokeGenerator();
        Scanner scanner = new Scanner(System.in);
        System.out.println("=== Random Joke Generator ===");
        while (true) {
            System.out.println("\n1. Get a random joke");
            System.out.println("2. Show all jokes");
            System.out.println("3. Add a joke");
            System.out.println("4. Remove a joke");
            System.out.println("5. Filter jokes by category");
            System.out.println("6. Show categories");
            System.out.println("7. Save jokes to file");
            System.out.println("8. Load jokes from file");
            System.out.println("9. Exit");
            System.out.print("Choose: ");
            String choice = scanner.nextLine().trim();
            switch (choice) {
                case "1":
                    displayJoke(gen.getRandomJoke());
                    break;
                case "2":
                    List<Joke> all = gen.getAllJokes();
                    if (all.isEmpty()) System.out.println("No jokes.");
                    else {
                        System.out.println("\nAll jokes:");
                        for (Joke j : all)
                            System.out.printf("[%d] %s (%s)\n", j.id, j.text, j.category);
                    }
                    break;
                case "3":
                    System.out.print("Enter your joke: ");
                    String text = scanner.nextLine().trim();
                    if (text.isEmpty()) { System.out.println("Joke cannot be empty."); break; }
                    System.out.print("Category (optional): ");
                    String cat = scanner.nextLine().trim();
                    Joke joke = gen.addJoke(text, cat);
                    System.out.printf("Joke added with ID %d.\n", joke.id);
                    break;
                case "4":
                    System.out.print("Enter joke ID to remove: ");
                    int id = scanner.nextInt(); scanner.nextLine();
                    if (gen.removeJoke(id)) System.out.println("Joke removed.");
                    else System.out.println("Joke not found.");
                    break;
                case "5":
                    System.out.print("Enter category: ");
                    String catFilter = scanner.nextLine().trim();
                    List<Joke> filtered = gen.getByCategory(catFilter);
                    if (filtered.isEmpty())
                        System.out.printf("No jokes in category '%s'.\n", catFilter);
                    else {
                        System.out.printf("\nJokes in category '%s':\n", catFilter);
                        for (Joke j : filtered)
                            System.out.printf("[%d] %s\n", j.id, j.text);
                    }
                    break;
                case "6":
                    List<String> cats = gen.getCategories();
                    if (cats.isEmpty()) System.out.println("No categories.");
                    else System.out.println("Categories: " + String.join(", ", cats));
                    break;
                case "7":
                    System.out.print("Filename: ");
                    String fname = scanner.nextLine().trim();
                    gen.saveToFile(fname);
                    System.out.printf("Saved to %s.\n", fname);
                    break;
                case "8":
                    System.out.print("Filename: ");
                    fname = scanner.nextLine().trim();
                    gen.loadFromFile(fname);
                    System.out.printf("Loaded from %s.\n", fname);
                    break;
                case "9":
                    System.out.println("Goodbye! 😄");
                    scanner.close();
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}

File: joke_generator.py
# joke_generator.py
import json
import random
import sys
from typing import List, Dict, Optional

class JokeGenerator:
    def __init__(self):
        self.jokes: List[Dict] = []
        self._load_default_jokes()

    def _load_default_jokes(self):
        self.jokes = [
            {"id": 1, "text": "Why don't scientists trust atoms? Because they make up everything!", "category": "science"},
            {"id": 2, "text": "What do you call a fake noodle? An impasta!", "category": "food"},
            {"id": 3, "text": "Why did the scarecrow win an award? Because he was outstanding in his field!", "category": "pun"},
            {"id": 4, "text": "I'm reading a book on anti-gravity. It's impossible to put down!", "category": "book"},
            {"id": 5, "text": "What's the best thing about Switzerland? I don't know, but the flag is a big plus!", "category": "geography"},
            {"id": 6, "text": "Why do programmers prefer dark mode? Because light attracts bugs!", "category": "programming"},
            {"id": 7, "text": "I told my wife she was drawing her eyebrows too high. She looked surprised.", "category": "marriage"},
            {"id": 8, "text": "What do you call a bear with no teeth? A gummy bear!", "category": "animal"},
            {"id": 9, "text": "Why don't eggs tell jokes? They'd crack each other up!", "category": "food"},
            {"id": 10, "text": "I used to play piano by ear, but now I use my hands.", "category": "music"},
        ]

    def get_random_joke(self) -> Optional[Dict]:
        return random.choice(self.jokes) if self.jokes else None

    def get_all_jokes(self) -> List[Dict]:
        return self.jokes

    def add_joke(self, text: str, category: str = "general") -> Dict:
        new_id = max([j["id"] for j in self.jokes]) + 1 if self.jokes else 1
        joke = {"id": new_id, "text": text, "category": category}
        self.jokes.append(joke)
        return joke

    def remove_joke(self, joke_id: int) -> bool:
        for i, j in enumerate(self.jokes):
            if j["id"] == joke_id:
                del self.jokes[i]
                return True
        return False

    def get_by_category(self, category: str) -> List[Dict]:
        return [j for j in self.jokes if j["category"].lower() == category.lower()]

    def get_categories(self) -> List[str]:
        return sorted(set(j["category"] for j in self.jokes))

    def save_to_file(self, filename: str):
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(self.jokes, f, indent=2, ensure_ascii=False)

    def load_from_file(self, filename: str):
        try:
            with open(filename, 'r', encoding='utf-8') as f:
                self.jokes = json.load(f)
        except FileNotFoundError:
            print(f"File {filename} not found.")

def display_joke(joke):
    print(f"\n😂 Joke: {joke['text']}")
    if joke.get('category'):
        print(f"📁 Category: {joke['category']}")

def main():
    gen = JokeGenerator()
    print("=== Random Joke Generator ===")
    while True:
        print("\n1. Get a random joke")
        print("2. Show all jokes")
        print("3. Add a joke")
        print("4. Remove a joke")
        print("5. Filter jokes by category")
        print("6. Show categories")
        print("7. Save jokes to file")
        print("8. Load jokes from file")
        print("9. Exit")
        choice = input("Choose: ").strip()
        if choice == "1":
            joke = gen.get_random_joke()
            if joke:
                display_joke(joke)
            else:
                print("No jokes available.")
        elif choice == "2":
            jokes = gen.get_all_jokes()
            if jokes:
                print("\nAll jokes:")
                for j in jokes:
                    print(f"[{j['id']}] {j['text']} ({j['category']})")
            else:
                print("No jokes.")
        elif choice == "3":
            text = input("Enter your joke: ").strip()
            if not text:
                print("Joke cannot be empty.")
                continue
            cat = input("Category (optional): ").strip() or "general"
            joke = gen.add_joke(text, cat)
            print(f"Joke added with ID {joke['id']}.")
        elif choice == "4":
            try:
                jid = int(input("Enter joke ID to remove: ").strip())
                if gen.remove_joke(jid):
                    print("Joke removed.")
                else:
                    print("Joke not found.")
            except ValueError:
                print("Invalid ID.")
        elif choice == "5":
            cat = input("Enter category: ").strip()
            jokes = gen.get_by_category(cat)
            if jokes:
                print(f"\nJokes in category '{cat}':")
                for j in jokes:
                    print(f"[{j['id']}] {j['text']}")
            else:
                print(f"No jokes in category '{cat}'.")
        elif choice == "6":
            cats = gen.get_categories()
            if cats:
                print("Categories:", ", ".join(cats))
            else:
                print("No categories.")
        elif choice == "7":
            fname = input("Filename: ").strip()
            gen.save_to_file(fname)
            print(f"Saved to {fname}.")
        elif choice == "8":
            fname = input("Filename: ").strip()
            gen.load_from_file(fname)
            print(f"Loaded from {fname}.")
        elif choice == "9":
            print("Goodbye! 😄")
            break
        else:
            print("Invalid choice.")

if __name__ == "__main__":
    main()
